hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/table/upgrade/TestUpgradeDowngrade.java

```diff
@@ -832,12 +875,87 @@ private void validateDataConsistency(Dataset<Row> originalData, HoodieTableMetaC
 
   private void performDataValidationOnTable(HoodieTableMetaClient metaClient, String stage) {
     LOG.info("Performing data validation on table {}", stage);
-    
+
     try {
       Dataset<Row> tableData = readTableData(metaClient, stage);
       LOG.info("Data validation passed {} (table accessible, {} rows)", stage, tableData.count());
     } catch (Exception e) {
       throw new RuntimeException("Data validation failed " + stage, e);
     }
   }
+
+  @ParameterizedTest
+  @MethodSource("testArgsPayloadUpgradeDowngrade")
+  public void testPayloadUpgradeDowngrade(String tableType, RecordMergeMode recordMergeMode, String payloadType) throws Exception {
+    LOG.info("Testing payload upgrade/downgrade for: {} (tableType: {}, recordMergeMode: {})",
+        payloadType, tableType, recordMergeMode);
+
+    // Load v6 fixture for this payload type
+    HoodieTableMetaClient metaClientV6 = loadPayloadFixtureTable(HoodieTableVersion.SIX, payloadType);
+    assertEquals(HoodieTableVersion.SIX, metaClientV6.getTableConfig().getTableVersion(),
+        "Fixture table should be at version 6 for payload: " + payloadType);
+
+    // Read original data before upgrade
+    Dataset<Row> originalDataV6 = readTableData(metaClientV6, "original v6 for " + payloadType);
+
+    HoodieWriteConfig config = createWriteConfig(metaClientV6, true);
+
+    // Test upgrade v6 -> v8
+    new UpgradeDowngrade(metaClientV6, config, context(), SparkUpgradeDowngradeHelper.getInstance())
+        .run(HoodieTableVersion.EIGHT, null);
+
+    HoodieTableMetaClient metaClientV8 = HoodieTableMetaClient.builder()
+        .setConf(storageConf().newInstance())
+        .setBasePath(metaClientV6.getBasePath())
+        .build();
+
+    assertEquals(HoodieTableVersion.EIGHT, metaClientV8.getTableConfig().getTableVersion(),
+        "Table should be upgraded to version 8 for payload: " + payloadType);
+    validateDataConsistency(originalDataV6, metaClientV8, "after v6->v8 upgrade for " + payloadType);
+
+    // Test upgrade v8 -> v9 (need to set record merge mode first)
+    Properties props = new Properties();
+    props.setProperty(HoodieTableConfig.RECORD_MERGE_MODE.key(), recordMergeMode.name());
```

Comment:
why merge mode is required here? can you shed some light

Reply:
This is a miss on my side and not needed, will remove